### ISSUE TITLE
training and validation images are same, but when training loss is different

### ISSUE THREAD

**Issue by DasolHan** [keras-team/keras#8643]
Hello,

I'm trying to test my model, but i have problem

i only use one image, i assign same image for training and validation 

after then, when i train with model.fit_generator, 

network produce different loss for training and validation losses

what happen?

**Comment by DasolHan** [comment]
plus, i don't use any dropout layer for my model

**Comment by teja-bogireddy** [comment]
With one image, the training loss be x. Based 'x' weights are updated. Now the same image is fed into the model with updated weight. Hence gives different loss.

**Comment by DasolHan** [comment]
when i debug loss without fit_generator, 
using output of model.predict to calculate loss,

the loss is same as last validation loss
is it correct?

and when i train the model, training loss is not decreasing, for example, 
"loss: 0.3111" 

but

validation loss is keep decreasing until "0.0xx"

**Comment by teja-bogireddy** [comment]
Are validation loss and training loss exactly same?
Or do they differ little?(0.01 to 0.05)

If they differ even a little,then my theory is correct.
Can you show your ImageGenerator ?

**Comment by DasolHan** [comment]
Actually i'm working on object detection and trying to implement yolo v2

as your request, upload my batch generator

from keras.utils import Sequence
from utils.box import Box, box_iou
import numpy as np
import cv2
import random
import copy


class BatchProvider(Sequence):
    def __init__(self,
                 config,
                 dataset,
                 anchors,
                 classes,
                 batch_size,
                 augmentation=True,
                 debug=False
                 ):
        self.config       = config
        self.dataset      = dataset
        self.anchors      = [Box(0, 0, anchors[i, 0], anchors[i, 1]) for i in range(len(anchors))]
        self.classes      = classes
        self.batch_size   = batch_size
        self.counter      = 0
        self.augmentation = augmentation
        self.debug        = debug

        # shuffle dataset
        keys                 = self.dataset.keys()
        self.shuffled_keys   = random.sample(keys, len(keys))

    def __len__(self):
        return int(len(self.dataset) / self.batch_size)

    def __getitem__(self, idx):

        first_index = idx * self.batch_size
        last_index  = (idx + 1) * self.batch_size

        if last_index > len(self.dataset):
            last_index  = len(self.dataset)
            first_index = last_index - self.batch_size

        # Create empty numpy for x_batch and y_batch
        x_batch = np.zeros((self.batch_size, self.config['IMG_INPUT_H'], self.config['IMG_INPUT_W'], 3))
        y_batch = np.zeros((self.batch_size,
                            int(self.config['IMG_INPUT_H'] / self.config['SHRINK_FACTOR']), int(self.config['IMG_INPUT_W'] / self.config['SHRINK_FACTOR']),
                            self.config['N_ANCHORS'], 5 + self.config['N_CLASSES']))

        for i, train_instance in enumerate(self.shuffled_keys[first_index:last_index]):

            filename    = train_instance
            objects     = self.dataset[filename]

            try:
                img = cv2.imread(filename)
                height, width, _ = img.shape

            except IOError:
                raise IOError("%s Check input filename." % filename)

            GRID_W = int(self.config['IMG_INPUT_W'] / self.config['SHRINK_FACTOR'])
            GRID_H = int(self.config['IMG_INPUT_H'] / self.config['SHRINK_FACTOR'])

            # augment input image and adjust objects's position and size
            if self.augmentation:
                if self.debug:
                    cv2.imshow('before augment', img)
                img, objects = augment_image(img, objects)


            if self.debug:
                print('[DEBUG] {} put into x_batch[{}]'.format(filename, i))
                gt_img = np.copy(img)

            aug_img    = preprocess_img(img)
            aug_img    = cv2.resize(aug_img, (self.config['IMG_INPUT_W'], self.config['IMG_INPUT_H']))

            x_batch[i] = aug_img

            for object in objects:
                bbox, label = object

                index = np.where(self.classes == label)[0][0]

                one_hot = np.eye(self.config['N_CLASSES'])[index]

                # convert integer value of bounding box to relative value
                cx, cy, w, h = bbox.to_relative_size((float(width), float(height)))

                # on grid coordinate
                grid_x = cx * GRID_W
                grid_y = cy * GRID_H

                grid_x = int(np.floor(grid_x))
                grid_y = int(np.floor(grid_y))

                if grid_x < GRID_W and grid_y < GRID_H:

                    # find the best anchor that fit to this object
                    best_anchor = -1
                    max_iou = -1

                    grid_w = bbox.w / width * GRID_W
                    grid_h = bbox.h / height * GRID_H

                    gt_bbox = Box(0, 0, grid_w, grid_h)

                    for j in range(len(self.anchors)):

                        anchor = self.anchors[j]
                        iou = box_iou(gt_bbox, anchor)

                        if max_iou < iou:
                            best_anchor = j
                            max_iou = iou

                    # assign ground truth cx, cy, w, h confidence and one-hot vector of class to y_batch
                    y_batch[i, grid_y, grid_x, best_anchor, 0:4] = [cx, cy, w, h]
                    y_batch[i, grid_y, grid_x, best_anchor, 4] = 1.
                    y_batch[i, grid_y, grid_x, best_anchor, 5:] = one_hot

                    if self.debug:
                        print('[DEBUG] gt:', grid_y, grid_x, best_anchor, [cx, cy, w, h])
                        #print('[DEBUG] Encoded:', y_batch[i, grid_y, grid_x, :, :])

                        # draw original box on the image
                        cord = bbox.to_opencv_format()
                        x1, y1 = cord[0][0]
                        x2, y2 = cord[0][1]
                        color = np.random.randint(0, 255, [3])
                        color = (int(color[0]), int(color[1]), int(color[2]))
                        # for whole object draw on final result image
                        cv2.rectangle(gt_img, (x1, y1), (x2, y2), color, 1)
                        cv2.putText(gt_img, label, (x1 + 2, y1 - 10), 0, 1.2e-3 * gt_img.shape[0], color, 1)
                        # for one object draw on image
                        anchor_img = np.copy(img)
                        cv2.rectangle(anchor_img, (x1, y1), (x2, y2), color, 1)
                        cv2.putText(anchor_img, label, (x1 + 2, y1 - 10), 0, 1.2e-3 * gt_img.shape[0], color, 1)
                        #cv2.imshow('object', anchor_img)
                        #cv2.waitKey(0)

            if self.debug:
                cv2.imshow('gt_image', gt_img)
                cv2.waitKey(0)

        # new batch created
        self.counter += 1

        return x_batch, y_batch.reshape([self.batch_size, int(self.config['IMG_INPUT_H'] / self.config['SHRINK_FACTOR']), int(self.config['IMG_INPUT_W'] / self.config['SHRINK_FACTOR']), self.config['N_ANCHORS'] * (5 + self.config['N_CLASSES'])])

    def on_epoch_end(self):
        self.counter = 0
        # shuffle dataset
        keys = self.dataset.keys()
        self.shuffled_keys = random.sample(keys, len(keys))


def preprocess_img(img):
    img = img / 255.

    return img

def augment_image(img, objects):
    aug_img      = np.copy(img)
    copy_objects = copy.deepcopy(objects)

    height, width, _ = img.shape

    # scale the image
    scale = np.random.uniform() / 10. + 1.
    aug_img = cv2.resize(aug_img, (0, 0), fx=scale, fy=scale)

    # translate the image
    max_offx = (scale - 1.) * width
    max_offy = (scale - 1.) * height
    offx = int(np.random.uniform() * max_offx)
    offy = int(np.random.uniform() * max_offy)
    aug_img = aug_img[offy: (offy + height), offx: (offx + width)]

    # changing brightness
    aug_img = change_brightness(aug_img)

    # flip the image
    p = np.random.uniform(0, 1)
    if p >= 0.5:
        aug_img = cv2.flip(aug_img, 1)

    # adjust obejct's position and size
    new_bbox = []
    for object in copy_objects:
        gt_box, gt_label = object
        cx, cy, w, h = gt_box.to_array()
        cx = int(cx * scale - offx)
        cx = max(min(cx, width), 0)
        cy = int(cy * scale - offy)
        cy = max(min(cy, height), 0)

        if p >= 0.5:
            cx = width - cx

        new_bbox.append([Box(cx, cy, w*scale, h*scale), gt_label])

    return aug_img, new_bbox


def change_brightness(img):
    img1 = cv2.cvtColor(img, cv2.COLOR_RGB2HSV)
    img1 = np.array(img1, dtype=np.float64)
    random_bright = 0.5 + np.random.uniform()
    img1[:, :, 2] = img1[:, :, 2] * random_bright
    img1[:, :, 2][img1[:, :, 2] > 255] = 255

    img1 = np.array(img1, dtype=np.uint8)
    img1 = cv2.cvtColor(img1, cv2.COLOR_HSV2RGB)

    return img1

**Comment by brucechou1983** [comment]
Same here. I use two ImageDataGenerator with the same seed and same parameters. The validation error is like 10x~100x the training error. Although the training error still goes down, it's very weird. I know that difference between these two errors is normal. But I thought 100x is a bit crazy.
